x/transferinject/keeper/keeper.go

```diff
@@ -0,0 +1,168 @@
+	if err != nil {
+		return 0, fmt.Errorf("cannot extract rollapp id from packet: %w", err)
+	}
+	// TODO: consider that other chains also might want this feature
+	if rollapp == nil {
+		return t.ICS4Wrapper.SendPacket(ctx, chanCap, destinationPort, destinationChannel, timeoutHeight, timeoutTimestamp, data)
+	}
+
+	packet := new(transfertypes.FungibleTokenPacketData)
+	if err = types.ModuleCdc.UnmarshalJSON(data, packet); err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet data: %s", err.Error())
+	}
+
+	// check if the rollapp already contains the denom metadata
+	if ContainsFunc(rollapp.TokenMetadata, func(dm *rtypes.TokenMetadata) bool { return dm.Base == packet.Denom }) {
+		return t.ICS4Wrapper.SendPacket(ctx, chanCap, destinationPort, destinationChannel, timeoutHeight, timeoutTimestamp, data)
+	}
+
+	denomMetadata, ok := t.bankKeeper.GetDenomMetaData(ctx, packet.Denom)
+	if !ok {
+		return 0, errorsmod.Wrapf(errortypes.ErrInvalidType, "denom metadata not found")
+	}
+
+	packet.Memo, err = types.AddDenomMetadataToMemo(packet.Memo, denomMetadata)
+	if err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrInvalidType, "cannot create injected value")
+	}
+
+	data, err = types.ModuleCdc.MarshalJSON(packet)
+	if err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrInvalidType, "cannot marshal ICS-20 transfer packet data")
+	}
+
+	return t.ICS4Wrapper.SendPacket(ctx, chanCap, destinationPort, destinationChannel, timeoutHeight, timeoutTimestamp, data)
+}
+
+// OnAcknowledgementPacket adds the token metadata to the rollapp if it doesn't exist
+func (t *Keeper) OnAcknowledgementPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+) error {
+	var data transfertypes.FungibleTokenPacketData
+	if err := types.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		return errorsmod.Wrapf(errortypes.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet data: %s", err.Error())
+	}
+
+	if len(data.Memo) == 0 {
+		return t.IBCModule.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	}
+
+	packetMetaData, err := types.ParsePacketMetadata(data.Memo)
+	if errors.Is(err, types.ErrMemoUnmarshal) || errors.Is(err, types.ErrMemoDMEmpty) {
+		return t.IBCModule.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	}
+	if err != nil {
+		return err
+	}
+
+	rollapp, err := t.rollappKeeper.ExtractRollappFromChannel(ctx, packet.SourcePort, packet.SourceChannel)
+	if err != nil {
+		return fmt.Errorf("cannot extract rollapp id from packet: %w", err)
+	}
+	if rollapp == nil {
+		return t.IBCModule.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	}
+
+	dm := packetMetaData.DenomMetadata
+
+	if ContainsFunc(rollapp.TokenMetadata, func(tm *rtypes.TokenMetadata) bool { return tm.Base == dm.Base }) {
+		return t.IBCModule.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	}
```

Comment:
can you add a a comment explaining why total equality is not used

Reply:
also add a comment explaining why it short circuits
(I understand why, but for future reading)